### PR TITLE
Refactor settings to AndroidX

### LINKED ISSUE #10579: Refactor settings to use PreferenceFragments instead of PreferenceActivity
## Help needed
While deep-diving into the theming issue (#9373) yesterday and playing around with different strategies, I've stumbled upon our outdated preferences implementation. We are still using `PreferenceActivity` with several preference screens, which is deprecated for a long time. What is worse is that this is technically not compatible to the rest of our app, which is using AndroidX and `AppCompatActivity` - which means a lot of extra work and inconsistencies.

As I have zero experience with fragment handling I'm asking for help, so if someone is willing to migrate our preferences from `PreferenceActivity` to `AppCompatActivity` and `PreferenceFragments` (or even some full up-to-date AndroidX way), please go ahead, as this will help us moving forward towards a consistent theming approach.

### PR DESCRIPTION
**Pull request by SchoolGuy**
## Description

- Changes to AndroidX settings
- Uses MaterialUI for the Settings
- Removes the custom Preference Implementation and replaces the content with defaults

## Related issues

Fixes #10579 

Related Board with sub-issues: https://github.com/cgeo/cgeo/projects/21

## Additional context

- This will PR will just get big, there is no away around it. I know it is ugly but I won't cut it into multiple.
- Once I am done I will force push everything, until then don't give a damn about the Git history please.

## Plan for this PR

- [x] Splitup the one big XML into smaller ones
- [x] Create all fragments and replace the TODOs with their names
- [x] ~~Obsolete all custom Preference implementations with default ones~~
- [ ] Move the custom code from the main SettingsActivity into the fragments
- [x] Fix the styling of the settings
- [ ] Fix all bugs we can find

### CODE REVIEW

**Comment by SchoolGuy** [comment]
CC @TacoTheDank here it is, not nearly ready except for your GH Inbox. :D

**Comment by SchoolGuy** [comment]
As a comparison here is the WhereYouGo PR: https://github.com/cgeo/WhereYouGo/pull/240

File: main/build.gradle [hunk 1]
```diff
@@ -335,6 +335,10 @@ dependencies {
     implementation "org.mapsforge:mapsforge-map-reader:$mapsforgeVersion"
     implementation "org.mapsforge:mapsforge-themes:$mapsforgeVersion"
 
+    // Settings Library + Appcompat for AndroidX
+    implementation 'androidx.appcompat:appcompat:1.2.0'
```
Comment (TacoTheDank) [inline main/build.gradle:339]: AndroidX AppCompat is already specified below at line 383 - this isn't needed.

**Comment by TacoTheDank** [comment]
Things look ok for now - only the XML changes are pushed ATM, so I can't say anything about the code changes.

File: main/build.gradle (hunk 1 above)
Comment (SchoolGuy) [reply main/build.gradle]: Fixed via 423e660

**Comment by SchoolGuy** [comment]
That's it for today. I will continue with replacing all custom Preference implementations when possible. Like in WhereYouGo I guess I can't mimic the exact behavior but I will do my best.

**Comment by eddiemuc** [comment]
ok to test

**Comment by SchoolGuy** [comment]
Okayyyy... This evening was fruitful in terms of knowledge but the rest we need to discuss now.

Our DialogPreference is for the garbage can because of the following nice lines of code: https://github.com/androidx/androidx/blob/8cb282ccdbb00687dbf253a4419ded0dfc786fb5/preference/preference/src/main/java/androidx/preference/PreferenceFragment.java#L605-L638

TLDR in case you are not interested in reading Jetpack Compose Code: We can't extend a `DialogFragment` anymore because they hardcode the list of `DialogFragments` available.

The only solution I currently know of: Use a "normall" Preference which has an Intent. That Intent then opens the Dialog which can get a SharedPreferenceManager via the Context and then we can set the value of the Preference again.

The next problem is the `AbstractInfoPreference`. I have not found a way yet to use `setWidgetLayoutResource`/`setLayoutResource` like the CheckboxPreference and inject a layout. The old way via the Inflater is not working anymore sadly. Maybe I can make it work but I couldn't so far.

I will try to continue tomorrow with at least one of the two problems. Once both problems are dealt with i can compile the app again and everything should be visible in read-only mode at least without problems.

**Comment by TacoTheDank** [comment]
@SchoolGuy cgeo doesn't use Compose though...? Am I missing something?

**Comment by SchoolGuy** [comment]
Update: So I have nasty merge conficts I won't resolve now. BUT its compiling again. Nothing is working again except scrolling through most of the screen but the rough idea of where we will go should be visible. I am throwing out major pieces of code which only differs in style to the default in very small things. I believe that if we really want to keep these minor adjustments then we need to readd them at the end. What I can say is that there is a lot of redundancy which I will be able to resolve. It takes long and I am very sorry for this but I havn't forgotten this PR!

**Comment by SchoolGuy** [comment]
So I didn't resolve it very nice but well it is resolved. Maybe I will redo it once I am done with the PR.

**Comment by fm-sys** [comment]
@Lineflyer `ok to test` doesn't seem to work for me and @eddiemuc (see #11697 as well). Can you accept both PR's and/or adapt CI permissions for us?

**Comment by Lineflyer** [comment]
add to whitelist

@fm-sys Not sure, what permissions are needed here. By now I added @SchoolGuy to our whitelist.

**Comment by bekuno** [comment]
> 
> 
> @Lineflyer `ok to test` doesn't seem to work for me and @eddiemuc (see #11697 as well). Can you accept both PR's and/or adapt CI permissions for us?

I added you both to the "GitHub Pull Request Builder" admin list at CI.

**Comment by moving-bits** [comment]
Thanks @SchoolGuy for diving into this huge task. To make it easier to keep track of things I have created a GitHub project within c:geo to collect all issues related to refactoring, see https://github.com/cgeo/cgeo/projects/21
If you like to, you can create separate issues for the tasks you need for refactoring, and assign the project "Refactor settings" to them. Also helpful for fine-tuning after a first merge.

**Comment by SchoolGuy** [comment]
Okay so I sat down in the morning and tried to get the reopening of the Fragments in the correct order (thus also the `redrawScreen`) method working. I also tried getting to work some of the details in the fragments.

A big problem in my eyes is that in the application most of the Settings and access to SharedPreferences is static (`Settings.java`) in WhereYouGo I had the same problem but due to the app being smaller it was easier to solve. The modularization and changed inheritance of the `SettingsActivity.java` kills most of the old ways of  arranging and modifying the UI. I will try to reimplement this via ViewModels and similar related mechanisms. This basically boils down to `findPreference` not being available in the `SettingsActivity.java` which leads to all the helper and generalized methods for creating the correct UI not working anymore.

Second problem I don't know yet how to solve it is that the avatar picture we download from the providers are not a drawable and thus I can't use `setIcon` to actually set the resource. I think this is easy to solve but I need to sleep some nights what the best approach is (I have ideas but nothing specific yet).

Another problem I encountered is that the Settings don't have E2E tests, so I can't verify what user interaction paths work and what not. Probably I need to create them after this PR is done.

For opening the correct Fragment after something has changed I believe I will choose [this](https://stackoverflow.com/a/36512646/4730773) approach. For all the other problems I don't have a solution yet in my head.

Once the application compiles again I will push the next commit. I hope that next weekend I will be able to put a full day into this PR so we get this PR in a kind of alpha state.

**Comment by fm-sys** [comment]
> A big problem in my eyes is that in the application most of the Settings and access to SharedPreferences is static (Settings.java)

Aren't the androidX preferences still compatible with `SharedPreferences`? What's the reason it does not work anymore? It would be very sad to lose this easy way of accessing user settings...

> Second problem I don't know yet how to solve it is that the avatar picture we download from the providers are not a drawable and thus I can't use setIcon to actually set the resource. 

The avatar handling will change a bit with  #10928, but in general I don't think that this feature has too much priority. In that PR I get the avatar drawable using this code:
```
final HtmlImage imgGetter = new HtmlImage(HtmlImage.SHARED, false, false, false);
AndroidRxUtils.andThenOnUi(AndroidRxUtils.networkScheduler,
    () -> imgGetter.getDrawable(Settings.getAvatarUrl((IAvatar) conn)),
    img -> View.setImageDrawable(img));

**Comment by Lineflyer** [comment]
retest this please

**Comment by SchoolGuy** [comment]
I will finish this next week as I finally will have some vacation then with time to program. If it isn't finished by then I will not be able to finish it by myself.

**Comment by SchoolGuy** [comment]
I rebased on todays `master` and will now start to fixup the bugs and missing features. After that some Git History cleanup is in order.

**Comment by SchoolGuy** [comment]
Okay there are currently two problems I have no idea of solving:

- How to fixup the `ColorpickerPreference`
- How to fixup the `TemplateTextPreference`

The latter one is already solved halfway through. Both custom preferences have the same problem: I don't find a way how to react to the `onBindDialogView` and similar events. If noone has an idea for me how to do this with the new androidx api then we need to change the whole implementation to an alert dialog. Since the preferences are using DialogFrames (which is elegant in my eyes), I would like to avoid that.

**Comment by moving-bits** [comment]
For the `ColorpickerPreference` I would propose the radical approach of skipping it for the time being. Users would have to edit color values using settings => system => view settings temporarily.

This would give us the chance to get this rather large PR into `master` a bit sooner, so that we can start fine-tuning. In parallel to that someone (probably me, as I had written the original `Colorpicker`) can start on a new approach to it, without blocking the whole migration.

**Comment by SchoolGuy** [comment]
@moving-bits and what about the `TemplateTextPreference`?

**Comment by moving-bits** [comment]
> @moving-bits and what about the `TemplateTextPreference`?

Don't know about the technical details, but can't we replace this with some button, opening a separate dialog? (Which might be a viable approach for `ColorPreference` as well, to be implemented later.)

Anyway, I guess the "skip it for now" approach may be applicable here as well, as template string probably do not change that frequently (and would still be available using settings => system => view settings).

**Comment by SchoolGuy** [comment]
@moving-bits Well in theory when we do this, we should be able to launch our own Activity and just pass back the results in an intent. We basically do this with opening the website already but this is a lot of effort compared to the few lines we have right now for those settings.

**Comment by SchoolGuy** [comment]
Tldr: I can make both work if we are fine with replacing the dialog with a dedicated page! But I didn't know if it is okay to change that behavior. If yes then I will include this with this PR.

**Comment by SchoolGuy** [comment]
@fm-sys @moving-bits I would love if you could test the PR and review the code I produced. I would love to get a TODO list which what I need to fix before we merge this.

**Comment by SchoolGuy** [comment]
Andddddd `onActivityResult` is deprecated: https://stackoverflow.com/a/63654043/4730773

I am touching one thing and three more pop up. This refactoring is getting out of hand and I don't know how to slice it into multiple. Please help guys.

**Comment by moving-bits** [comment]
> Please help guys.

Sorry, can't help currently. No dev PC at hand, and will be mostly AFK for the next week.

**Comment by eddiemuc** [comment]
> Andddddd `onActivityResult` is deprecated: https://stackoverflow.com/a/63654043/4730773
> 
> I am touching one thing and three more pop up. This refactoring is getting out of hand and I don't know how to slice it into multiple. Please help guys.

I strongly suggest to keep deprecated `onActivityResult` for now. This will be a whole topic on its own since we use it across nearly all activities today.

**Comment by SchoolGuy** [comment]
@eddiemuc Okay. Would you be able to create a list of things to fix before we merge? So bugs and currently not implemented/missing features?

**Comment by eddiemuc** [comment]
> @eddiemuc Okay. Would you be able to create a list of things to fix before we merge? So bugs and currently not implemented/missing features?

Hi @SchoolGuy I am not able to create such a list to be honest. I did not follow the topic previously. I would say that as soon as a working state is reached where code is already improved and all features still work (except for already agreed exceptions) then we should merge.

**Comment by SchoolGuy** [comment]
Back button was fixed thanks to this tutorial: https://blog.mindorks.com/implementing-android-jetpack-preferences

**Comment by SchoolGuy** [comment]
Okay navigation should work exactly like before. Now I need to check all settings manually if they work as intended. Sadly we don't have tests (definitely a task for a different PR), so I need to check everything manually.

**Comment by fm-sys** [comment]
@SchoolGuy I gave it a try in the emulator (Android 10). First of all, thanks for your work! Looks really promising already...

Let's create a list of bugs I have discovered: (probably far from being complete, but at least it's a good starting point)
- c:geo crashes when opening the map preferences
 ```
E/AndroidRuntime: FATAL EXCEPTION: main
    Process: cgeo.geocaching, PID: 25841
    java.lang.ClassCastException: java.lang.Integer cannot be cast to java.util.Set
        at android.app.SharedPreferencesImpl.getStringSet(SharedPreferencesImpl.java:298)
        at androidx.preference.Preference.getPersistedStringSet(Preference.java:1739)
        at androidx.preference.MultiSelectListPreference.onSetInitialValue(MultiSelectListPreference.java:205)
        at androidx.preference.Preference.onSetInitialValue(Preference.java:1614)
        at androidx.preference.Preference.dispatchSetInitialValue(Preference.java:1587)
        at androidx.preference.Preference.onAttachedToHierarchy(Preference.java:1311)
        at androidx.preference.Preference.onAttachedToHierarchy(Preference.java:1326)
        at androidx.preference.PreferenceGroup.addPreference(PreferenceGroup.java:249)
        at androidx.preference.PreferenceGroup.addItemFromInflater(PreferenceGroup.java:170)
        at androidx.preference.PreferenceInflater.rInflate(PreferenceInflater.java:345)
        at androidx.preference.PreferenceInflater.rInflate(PreferenceInflater.java:346)
        at androidx.preference.PreferenceInflater.inflate(PreferenceInflater.java:157)
        at androidx.preference.PreferenceInflater.inflate(PreferenceInflater.java:109)
        at androidx.preference.PreferenceManager.inflateFromResource(PreferenceManager.java:216)
        at androidx.preference.PreferenceFragmentCompat.setPreferencesFromResource(PreferenceFragmentCompat.java:377)
        at cgeo.geocaching.settings.fragments.PreferenceMapFragment.onCreatePreferences(PreferenceMapFragment.java:24)
        at androidx.preference.PreferenceFragmentCompat.onCreate(PreferenceFragmentCompat.java:160)
        at androidx.fragment.app.Fragment.performCreate(Fragment.java:2949)
        at androidx.fragment.app.FragmentStateManager.create(FragmentStateManager.java:475)
        at androidx.fragment.app.FragmentStateManager.moveToExpectedState(FragmentStateManager.java:278)
        at androidx.fragment.app.FragmentManager.executeOpsTogether(FragmentManager.java:2189)
        at androidx.fragment.app.FragmentManager.removeRedundantOperationsAndExecute(FragmentManager.java:2100)
        at androidx.fragment.app.FragmentManager.execPendingActions(FragmentManager.java:2002)
        at androidx.fragment.app.FragmentManager$5.run(FragmentManager.java:524)
        at android.os.Handler.handleCallback(Handler.java:883)
        at android.os.Handler.dispatchMessage(Handler.java:100)
        at android.os.Looper.loop(Looper.java:214)
        at android.app.ActivityThread.main(ActivityThread.java:7356)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:492)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:930)
```
- c:geo should restart when leaving the settings if `Appearance > Select Language` changed
- changing the `Appearance > Theme` setting looks weird. Probably a `requireActivity().recreate()` would do the trick instead of starting a new settings activity
- clicking on the connector name should launch the settings at sub-page `Services`
grafik
- same applies for `Backup/Restore`
grafik
- it would be nice, if it's still shown which connectors are active (show "Active" as summary)
grafik
- show current signature as summary
grafik
- allow multi-line text would help here (applies for all headings)
grafik
- crash when selecting a navigation tool `Routing/Navigation > Default/Secondary Navigation`
```
E/AndroidRuntime: FATAL EXCEPTION: main
    Process: cgeo.geocaching, PID: 27082
    java.lang.IllegalStateException: ListPreference requires an entries array and an entryValues array.
        at androidx.preference.ListPreferenceDialogFragmentCompat.onCreate(ListPreferenceDialogFragmentCompat.java:53)
        at androidx.fragment.app.Fragment.performCreate(Fragment.java:2949)
        at androidx.fragment.app.FragmentStateManager.create(FragmentStateManager.java:475)
        at androidx.fragment.app.FragmentStateManager.moveToExpectedState(FragmentStateManager.java:278)
        at androidx.fragment.app.FragmentManager.executeOpsTogether(FragmentManager.java:2189)
        at androidx.fragment.app.FragmentManager.removeRedundantOperationsAndExecute(FragmentManager.java:2100)
        at androidx.fragment.app.FragmentManager.execPendingActions(FragmentManager.java:2002)
        at androidx.fragment.app.FragmentManager$5.run(FragmentManager.java:524)
        at android.os.Handler.handleCallback(Handler.java:883)
        at android.os.Handler.dispatchMessage(Handler.java:100)
        at android.os.Looper.loop(Looper.java:214)
        at android.app.ActivityThread.main(ActivityThread.java:7356)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:492)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:930)
```
- everything is grayed out (`Routing/Navigation > Navigation Menu`)
grafik
- crash when trying to change routing profile (`Routing/Navigation > "Walk/Bike/Car" profile`)
```
E/AndroidRuntime: FATAL EXCEPTION: main
    Process: cgeo.geocaching, PID: 27415
    java.lang.IllegalStateException: ListPreference requires an entries array and an entryValues array.
        at androidx.preference.ListPreferenceDialogFragmentCompat.onCreate(ListPreferenceDialogFragmentCompat.java:53)
        at androidx.fragment.app.Fragment.performCreate(Fragment.java:2949)
        at androidx.fragment.app.FragmentStateManager.create(FragmentStateManager.java:475)
        at androidx.fragment.app.FragmentStateManager.moveToExpectedState(FragmentStateManager.java:278)
        at androidx.fragment.app.FragmentManager.executeOpsTogether(FragmentManager.java:2189)
        at androidx.fragment.app.FragmentManager.removeRedundantOperationsAndExecute(FragmentManager.java:2100)
        at androidx.fragment.app.FragmentManager.execPendingActions(FragmentManager.java:2002)
        at androidx.fragment.app.FragmentManager$5.run(FragmentManager.java:524)
        at android.os.Handler.handleCallback(Handler.java:883)
        at android.os.Handler.dispatchMessage(Handler.java:100)
        at android.os.Looper.loop(Looper.java:214)
        at android.app.ActivityThread.main(ActivityThread.java:7356)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:492)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:930)
```
- setting seems to have no effect:
grafik
- "start backup" has no effect
grafik

- using `app:iconSpaceReserved="false"` would make sense, since we don't have any icons on sub-pages

**Comment by SchoolGuy** [comment]
@fm-sys Thanks a lot! I will address all of this once I recovered from my physical back incident. Doctors said that starting on Saturday I can start working actively again on the computer.

**Comment by fm-sys** [comment]
At least for me, it would be ok to merge this PR even if it still contains minor regressions. 
These bugs should definitely be fixed before a release, but would also be ok to fix them on master because we have _much_ time and want to include this change into the next release anyway. 
The past has shown, that it is much easier to cooperate on such topics if the code base is contained on `master` and actively tested in the nightlys (build from `master` branch)

However, the Codacy warnings should be fixed before a merge, to stay aligned with cgeo's coding conventions. 

Furthermore, I have seen some old code pieces which are currently simply commented out. I guess, these are places where previous functionality is currently missing, so these places should either be tracked in a todo-list or implemented/removed/rewritten.

**Comment by fm-sys** [comment]
Oh, and of course I hope you get well soon!

**Comment by eddiemuc** [comment]
I full agree with @fm-sys . Let's get this large PR into master as soon as possible (means: checks are green) and fix remaining problems in collaboration.

@SchoolGuy also from my side: hope you get better soon. And thanks for all the contribution work!

**Comment by moving-bits** [comment]
I'm back from my business trip, can start helping to contribute on this. But I agree, that getting this into `master` first would make things easier. Although the Codacy issues should be fixed first.

@SchoolGuy 
Hope that you are feeling better meanwhile!
Do you want to fix the Codacy issues yourself, or shall I help you with them, adding commits to your branch?

**Comment by SchoolGuy** [comment]
@moving-bits Thanks that would indeed help. My back is healing slower then I personally have hoped for and long sitting sessions are still hard for me. I will continue to work on this as soon as health allows.

**Comment by moving-bits** [comment]
Ok, have added a commit to fix most Codacy / CheckStyle issues, all except the "unused local variable" / "unused method parameter" types, as those will probably be filled during upcoming work. Will do some testing as time permits, and create issues in the [project](https://github.com/cgeo/cgeo/projects/21) for my findings.

**Comment by fm-sys** [comment]
> Will do some testing as time permits, and create issues in the project for my findings.

@moving-bits just in case you have missed the post: I already have created a rather large list of issues some comments above, https://github.com/cgeo/cgeo/pull/11298#issuecomment-966393487

Not sure whether you want to create separate issues for all items, or move everything to ome ToDo list...

**Comment by moving-bits** [comment]
> @moving-bits just in case you have missed the post: I already have created a rather large list of issues some comments above, [#11298 (comment)](https://github.com/cgeo/cgeo/pull/11298#issuecomment-966393487)

Oh, now as you mention it I remember having seen the list, but admittedly have not cross-checked yesterday, but...

> Not sure whether you want to create separate issues for all items, or move everything to one ToDo list...

... I would prefer separate issues for things found in the new implementation, which gives more space for describing the specific problem, and, more important, makes tracking easier. I had created a dedicated project for tracking those issues some time ago.

**Comment by fm-sys** [comment]
> > Not sure whether you want to create separate issues for all items, or move everything to one ToDo list...
> 
> ... I would prefer separate issues for things found in the new implementation, which gives more space for describing the specific problem, and, more important, makes tracking easier. I had created a dedicated project for tracking those issues some time ago.

Ok, I've just created separate issues for these findings...

**Comment by SchoolGuy** [comment]
Okay we have a merge conflict now and it appears things have moved faster then I anticipated in the last days. While getting slowly back to my desk I will try to fixup as much of the separate issues as possible.

**Comment by SchoolGuy** [comment]
To summarize in a very direct way, what I thought I had expressed before: Many of those bugs exist because I didn't implement the functionality. I will start doing so soon. I will also try to comment on the individual bugs once I am in a position to fix them.

**Comment by SchoolGuy** [comment]
Merge conflict is magically gone! Yes! 😄

**Comment by moving-bits** [comment]
Regarding the Codacy issues: Most of them can be checked from within Android Studio, just install the CheckStyle extension...

image

... and run it before committing your changes - saves you some round trips:

image

**Comment by bekuno** [comment]
... and in the commit dialog activate "Scan with Checkstyle"
image

**Comment by SchoolGuy** [comment]
@moving-bits @bekuno I know but for now I ignore them because the parts which it complains about are not done yet.

**Comment by SchoolGuy** [comment]
I did rebase onto `master` so the conflict is gone.

@fm-sys @bekuno @moving-bits Please check my progress and see if we can close the issues I linked in the commits. I am aware that I haven't fixed all of the things you requested. The Git History is not yet fixed and I will do so once I have the code in a satisfying state.

**Comment by moving-bits** [comment]
Thanks @SchoolGuy for working on the issues. Out of the issues reported by me, #12108 is positively influenced (restart after changing language) - I've updated the issue accordingly.

Regarding collaboration: What would you prefer? - Keep away from settings and found issues regarding settings until you give a "go" (to not break anything you are currently working on), or shall we dive into findings as well and push commits to your branch? (Both would be totally ok for me.)

If the latter, it would help if you could use the project we have created for refactoring settings and move the items you are working on to the "in progress" lane to not start working on things in parallel, see https://github.com/cgeo/cgeo/projects/21

**Comment by SchoolGuy** [comment]
@moving-bits Since you can open PRs to my branch I would prefer either that or you just push directly onto it. As long as you don't force push you can break nothing. Since I can't move issues and don't have any special rights in the org/repo I can't do most of the things regarding issue management you proposed.

**Comment by SchoolGuy** [comment]
It appears that the Jenkins Build is broken... Can we do something about this?

**Comment by moving-bits** [comment]
> It appears that the Jenkins Build is broken... Can we do something about this?

@mucek4 is already working on it. But the cause for the failing builds seems to be a mystery still :-(

**Comment by moving-bits** [comment]
@SchoolGuy 
Regarding the button handling I introduced a little helper method `SettingsUtils.setPrefClick` which sets a `onPreferenceClickListener` on the preference identified by its resource id, and calls the given method when clicked.
This made re-enabling the buttons in Settings => System a one-liner. Will try for other buttons as well as time permits.

**Comment by SchoolGuy** [comment]
@moving-bits Thanks a lot! I will continue to work on Monday after work. I will try to reimplement the fragment based approach for opening the right page to fix the shortcuts.

**Comment by SchoolGuy** [comment]
Okay worst case approach is we do this: https://stackoverflow.com/questions/40679339/get-all-preferences-from-preferencefragment

Tldr: We loop over all fragments and then open the one where the preference we request is on. Let's see if I can find a better way tomorrow.

**Comment by kumy** [comment]
retest this

**Comment by kumy** [comment]
retest this

**Comment by SchoolGuy** [comment]
@moving-bits I am no maintainer of c:geo but I would vote for the following now:

- Fix Codacy up (if my last commit didn't do that already)
- Merge this PR and close issues which we fixed
- Create a new PR with E2E tests for the settings
- Discuss and fix the bugs in `master` via individual PRs for individual issues
- Release Alpha/Beta once we are certain that most things should work as expected

**Comment by SchoolGuy** [comment]
My personal lesson learnt on this PR is the following:

- Every customization away from the standard style guide is code to maintain and painful to port. The WhereYouGo port was much easier because there was way less customization in the settings.
- Many of your abstraction layers are very helpful and are a great piece of engineering. Others however literally cost me weeks and were just not needed in the end.
- While in WhereYouGo the Fragment based approach saved us a lot of lines to maintain, it seems that in c:geo it is a plus/minus zero PR roughly. Still I like the new approach as we now have a big monolithic class less.
- I am sadly a too big fan of Open Source. I have contributed to too many projects and my main project is https://github.com/cobbler/cobbler and I did not maintain it enough while working on this PR. So although I want to work more with you guys, I have to sadly say that I can't become a regular contributor here. I will still try to help out and finish the settings up and I will of course create PRs for bugs I have caused.

**Comment by moving-bits** [comment]
A big thank you @SchoolGuy for deep-diving into the issue of settings' migration and all the work you've put in so far! IMHO the current state, although not yet finished, is quite usable already, and further collaboration will be way easier after merge to `master`. Therefore I heartily support your proposal to merge this large PR.

Will do so tonight, if no one stops me :-)

> @moving-bits I am no maintainer of c:geo but I would vote for the following now:
> * Fix Codacy up (if my last commit didn't do that already)

There's one minor issue marked by Coday, an empty `if` statement. I guess this is ongoing work we will hit anyway on implementing the missing pieces, so no blocker, IMHO.

> * Merge this PR and close issues which we fixed

There's a merge conflict in `preferences.xml`, probably due to #12166, which removes the `pref_mapAutoDownloads` preference and a couple of dependencies. @SchoolGuy: Can you have a look into it?

> * Create a new PR with E2E tests for the settings
> * Discuss and fix the bugs in `master` via individual PRs for individual issues

A couple of issues are already fixed, and the related issues will be closed on merging this PR to `master`. I've updated the [project](https://github.com/cgeo/cgeo/projects/21) to reflect the implementation state.

> * Release Alpha/Beta once we are certain that most things should work as expected

Merging this to `master` will put it into our `alpha` channel immediately, which is our nightly versions. But as those are our most experienced users only, it's ok that not all settings are fully implemented already. For a `beta` most/all settings should be fully implemented, but as we have just released a new major version, we have plenty of time.

Again, highly appreciated that you've put all that energy and endurance into it!

**Comment by SchoolGuy** [comment]
@moving-bits If you don't have anything pending I will fix the Codacy and Merge conflict later in the afternoon!

**Comment by moving-bits** [comment]
> @moving-bits If you don't have anything pending I will fix the Codacy and Merge conflict later in the afternoon!

No pending code currently over here, thanks for fixing.

**Comment by SchoolGuy** [comment]
@moving-bits Done. I also replaced the "Tmp" commit with a littlebit better commit message.

**Comment by kumy** [comment]
rebuild this

**Comment by kumy** [comment]
retest this

**Comment by SchoolGuy** [comment]
It appears that we still have build and Codacy problems. :( I'll try to fix them either on Tuesday or Thursday.

**Comment by Lineflyer** [comment]
I was asked by @moving-bits if it would be OK to merge this PR to `master` to make it easier to continue bugfixing of this rather large change.

The minimum precondition I would see is, that all white-box text (Codacy, Checkstyle, etc.) errors are fixed. Furthermore ideally the functionality should be OK, but AFAIK there are still some known regressions (e.g. not possible to set paths). 

On the other side it would be easier to continue working on it even in the current state once its merged, as
a) More testers in nightly,
b) No more need for constant rebasing to master

So I would like to use our emoticon-feedback system :) to ask for the preference the @cgeo/team has:
Thumbs up = Merge to `master` as soon as tests are green, accept temporary regressions or missing functionality in the nightlies and continue fixing those in follow-up PRs
Thumbs down = Wait until all known regressions / missing functionality is added and only merge to `master` once its considered done completely

**Comment by SchoolGuy** [comment]
@baiti I will help you with rebasing. Sorry for the trouble.

**Comment by SchoolGuy** [comment]
Currently the test "GeokretyConnectorTest" is failing. I havn't investigated why but I will do so as soon as possible.

**Comment by SchoolGuy** [comment]
What is our state here now? Do we merge asap or should I drag this along further? If we drag it along, I would prefer to merge the other PRs which are dependent and I just rebase. It is a pita but I do get paid at work to rebase most of the time (ah what fun it is 😂 ) so it is less pita for me then for others I believe.

**Comment by moving-bits** [comment]
retest this please

**Comment by moving-bits** [comment]
Build is green, voting was 4:2 pro "merge now", so let's merge here to get feedback via the nightlies and for easier collaboration and issue tracking.
